runtime/master/src/main/java/org/apache/nemo/runtime/master/scheduler/SimulationScheduler.java

```diff
@@ -0,0 +1,579 @@
+    }
+
+    // Invoke doSchedule()
+    switch (newState) {
+      case COMPLETE:
+      case ON_HOLD:
+        // If the stage has completed
+        final String stageIdForTaskUponCompletion = RuntimeIdManager.getStageIdFromTaskId(taskId);
+        if (planStateManager.getStageState(stageIdForTaskUponCompletion).equals(StageState.State.COMPLETE)
+          && !planStateManager.isPlanDone()) {
+          doSchedule();
+        }
+        break;
+      case SHOULD_RETRY:
+        // Do retry
+        doSchedule();
+        break;
+      default:
+        break;
+    }
+
+    // Invoke taskDispatcher.onExecutorSlotAvailable()
+    switch (newState) {
+      // These three states mean that a slot is made available.
+      case COMPLETE:
+      case ON_HOLD:
+      case SHOULD_RETRY:
+        taskDispatcher.onExecutorSlotAvailable();
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * Handle metric message, as it would have been sent from org.apache.nemo.runtime.executor.MetricManagerWorker#send.
+   * @param metricType  type of metric.
+   * @param metricId    id of metric.
+   * @param metricField field of metric.
+   * @param metricValue value of metric.
+   */
+  void handleMetricMessage(final String metricType, final String metricId,
+                           final String metricField, final byte[] metricValue) {
+    final Class<Metric> metricClass = metricStore.getMetricClassByName(metricType);
+    // process metric message
+    try {
+      if (metricStore.getOrCreateMetric(metricClass, metricId).processMetricMessage(metricField, metricValue)) {
+        metricStore.triggerBroadcast(metricClass, metricId);
+      }
+    } catch (final Exception e) {
+      LOG.warn("Error when processing metric message for {}, {}, {}.", metricType, metricId, metricField);
+    }
+  }
+
+  /**
+   * The endpoint of the simulator. Collect the metric store, and terminate the simulator.
+   * @return the metrics of the simulation.
+   */
+  public MetricStore collectMetricStore() {
+    try {
+      // wait for metric flush
+      if (!metricCountDownLatch.await(10000, TimeUnit.MILLISECONDS)) {
+        LOG.warn("Terminating master before all executor terminated messages arrived.");
+      }
+    } catch (final InterruptedException e) {
+      LOG.warn("Waiting executor terminating process interrupted: ", e);
+      // clean up state...
+      Thread.currentThread().interrupt();
+    }
+
+    final MetricStore res = this.metricStore;
+    this.reset();
+    return res;
+  }
+
+  @Override
+  public void onSpeculativeExecutionCheck() {
+    // we don't simulate speculate execution yet.
+    return;
+  }
+
+  @Override
+  public void terminate() {
+    this.taskDispatcher.terminate();
+    this.executorRegistry.terminate();
+  }
+
+  /**
+   * Evaluator ActiveContext for the Simulation.
+   */
+  private static final class SimulationEvaluatorActiveContext implements ActiveContext {
```

Comment:
Would it be possible to not introduce this?

Reply:
I'm afraid that the active context is required to create executor representers that simulates executors